main.c

```diff
@@ -1,11 +1,103 @@
 #include <stdbool.h>
 #include <stdio.h>
 #include <string.h>
+#include <unistd.h>
 #include "mzcc.h"
 
+char *outfile = NULL, *infile = NULL;
+extern FILE *outfp;
+bool dump_ast;
+
+static void usage()
+{
+    fprintf(
+        stdout,
+        "\n"
+        "Usage : mzcc [ input file | - ]\n"
+        "\n"
+        "  -o filename            Output to the specified file\n"
+        "  --dump-ast             Print Abstract Syntax Tree\n"
+        "  [ input file | - ]     Specifying a file or use standard input\n\n");
+}
+
+void print_usage_and_exit()
+{
+    usage();
+    exit(1);
+}
+
+static void parse_args(int argc, char **argv)
+{
+    if (argc < 2) {
+        print_usage_and_exit();
+    }
+
+    while (true) {
+        argc--;
+        argv++;
+        if (!argc) {
+            break;
+        }
+
+        if ((*argv)[0] == '-') {
+            switch ((*argv)[1]) {
+            case '\0':
+                infile = "/dev/stdin";
+                break;
+            case 'o':
+                argc--;
+                argv++;
+                outfile = *argv;
+                break;
+            case '-':
+                if (!strcmp(*argv, "--dump-ast")) {
+                    dump_ast = true;
+                    break;
+                }
+            default:
+                print_usage_and_exit();
+            }
+        } else {
+            if (infile) {
+                // The second non-option argument is not what we expect.
+                print_usage_and_exit();
+            }
+            infile = argv[0];
+        }
+    }
+}
+
+static FILE *open_output_file()
+{
+    FILE *tmp = stdout;
+    if (outfile && !(tmp = fopen(outfile, "w"))) {
+        printf("Can not open file %s\n", outfile);
+        exit(1);
+    }
+    return tmp;
+}
+
+static FILE *open_input_file()
+{
+    if (!infile) {
+        printf("Input file is not specified\n\n");
+        print_usage_and_exit();
+    }
+
+    FILE *tmp = fopen(infile, "r");
+    if (tmp == NULL) {
+        printf("Can not open file %s\n", infile);
+        exit(1);
+    }
+    return tmp;
+}
+
 int main(int argc, char **argv)
 {
-    bool dump_ast = ((argc > 1) && !strcmp(argv[1], "--dump-ast"));
+    parse_args(argc, argv);
+    stdin = open_input_file();
```

Comment:
Avoid using variable name "stdin", which might cause confusion.

Reply:
I change the function like this :

```c
static void open_input_file()
{
    if (!infile) {
        printf("Input file is not specified\n\n");
        print_usage_and_exit();
    }
    
    if (!freopen(infile, "r", stdin)) {
        printf("Can not open file %s\n", infile);
        exit(1);
    }
}
```
Is that better ?